### PR TITLE
Added ConvertCommand with tests

### LINKED ISSUE #34: As a student who is going for exchange, I want to make sure that I bring enough foreign currency
... so that I do not have to use my credit card.

### LINKED ISSUE #66: Ang Chin Guan Melvin: convert command
v1.3: Can convert the currency of a `Transaction` or a specified amount.

### LINKED ISSUE #130: Exchange class to handle Currency conversion
Make a utility class to handle currency conversion for the `convert` command.

### PR DESCRIPTION
**Pull request by krusagiz**
Added ConvertCommand

Tests are included.

Resolves #34 Resolves #66 Resolves #130

### CODE REVIEW

**Review by lightz96 — CHANGES_REQUESTED** [review]
Unable to run the application. You should allow the creation of the JSON file if JSON file is missing.

**Comment by krusagiz** [comment]
> 
> 
> Unable to run the application. You should allow the creation of the JSON file if JSON file is missing.

This problem has been resolved. It now runs even without the JSON file, does not write a new currency file, however.

File: src/main/java/thrift/logic/commands/ConvertCommand.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof ConvertCommand // instanceof handles nulls
+                && amount == (((ConvertCommand) other).amount)
+                && currencies.equals(((ConvertCommand) other).currencies));
+    }
+
+    /**
+     * Generates the result message for conversions using instance variables
+     * @return Result message for conversions
```
Comment (lightz96) [inline src/main/java/thrift/logic/commands/ConvertCommand.java:66]: It should have an empty line between description and parameter section

File: src/main/java/thrift/logic/commands/ConvertCommand.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    private String generateConvertResult() {
+        return generateConvertResult(amount, currencies);
+    }
+
+    /**
+     * Generates the result message for conversions
+     * @return Result message for conversions
```
Comment (lightz96) [inline src/main/java/thrift/logic/commands/ConvertCommand.java:75]: It should have an empty line between description and parameter section.

File: src/main/java/thrift/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -137,4 +141,26 @@ public static Tag parseTag(String tag) throws ParseException {
         }
         return tagSet;
     }
+
+    /**
+     * Checks if the currencies are valid
+     * @param currencies The currencies
```
Comment (lightz96) [inline src/main/java/thrift/logic/parser/ParserUtil.java:148]: It should have an empty line between description and parameter section.

File: src/main/java/thrift/model/util/CurrencyUtil.java [hunk 4]
```diff
@@ -36,8 +73,9 @@ public static double convertFromDefaultCurrency(Map<String, Double> currencyMapp
         requireNonNull(currencyMappings);
         requireNonNull(value);
         requireNonNull(currency);
-        if (currencyMappings.containsKey(currency)) {
-            return (value * currencyMappings.get(currency));
+        if (currencyMappings.containsKey(currency.toUpperCase())) {
+            return (value * currencyMappings.get(currency.toUpperCase()));
+            //return value;
```
Comment (lightz96) [inline src/main/java/thrift/model/util/CurrencyUtil.java]: Please remove this.

File: src/main/java/thrift/model/util/CurrencyUtil.java [hunk 5]
```diff
@@ -57,11 +95,32 @@ public static double convertToDefaultCurrency(Map<String, Double> currencyMappin
...
+        if (currencyMappings.containsKey(currency.toUpperCase())) {
+            return (value / currencyMappings.get(currency.toUpperCase()));
         } else {
             return value;
         }
     }
 
+    /**
+     * Converts input currency amount from {@code currencyFrom} to {@code currencyTo} via {@link Value#DEFAULT_CURRENCY}
+     * @param currencyMappings Mapping of currency rates scaling from
```
Comment (lightz96) [inline src/main/java/thrift/model/util/CurrencyUtil.java:107]: It should have an empty line between description and parameter section.

**Review by lightz96 — CHANGES_REQUESTED** [review]
LGTM. This PR can merge once minor coding standards issues are fixed.

File: src/main/java/thrift/logic/commands/ConvertCommand.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        convertResultMsgSb.append(String.format(MESSAGE_BASE_CURRENCY_FORMAT, baseCurrency, amount));
+
+        for (int i = 1; i < currencies.size(); i++) {
+            Map<String, Double> currencyMap = CurrencyUtil.getCurrencyMap();
+
+            String targetCurrency = currencies.get(i);
+            double convertedValue = CurrencyUtil.convert(currencyMap, amount, baseCurrency, targetCurrency);
+
+            convertResultMsgSb.append(String.format(MESSAGE_TARGET_CURRENCY_FORMAT, targetCurrency, convertedValue));
```
Comment (limerencee) [inline src/main/java/thrift/logic/commands/ConvertCommand.java]: You could format the displayed `convertedValue` by using `DecimalFormatter` defined in `Value` class, like so:
`DECIMAL_FORMATTER.format(convertedValue)` <-- this returns a formatted `String`

File: src/main/java/thrift/logic/parser/CliSyntax.java [hunk 7]
```diff
@@ -13,4 +13,6 @@
     public static final Prefix PREFIX_REMARK = new Prefix("r/");
     public static final Prefix PREFIX_TAG = new Prefix("t/");
     public static final Prefix PREFIX_VALUE = new Prefix("v/");
+    public static final Prefix PREFIX_CURRENCY = new Prefix("c/");
```
Comment (limerencee) [inline src/main/java/thrift/logic/parser/CliSyntax.java]: Maybe we should keep this list sorted in alphabetical form?

File: src/main/java/thrift/logic/commands/ConvertCommand.java (hunk 1 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/commands/ConvertCommand.java:66]: Resolved.

**Review by limerencee — CHANGES_REQUESTED** [review]
LGTM! Awesome feature :smile:

File: src/main/java/thrift/logic/parser/ParserUtil.java (hunk 3 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/parser/ParserUtil.java:148]: Resolved.

File: src/main/java/thrift/model/util/CurrencyUtil.java (hunk 4 above)
Comment (krusagiz) [reply src/main/java/thrift/model/util/CurrencyUtil.java]: Removed.

File: src/main/java/thrift/logic/commands/ConvertCommand.java (hunk 2 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/commands/ConvertCommand.java:75]: Resolved

File: src/main/java/thrift/logic/commands/ConvertCommand.java (hunk 6 above)
Comment (limerencee) [reply src/main/java/thrift/logic/commands/ConvertCommand.java]: The `amount` variable above should also be formatted if that's the case, and `MESSAGE_BASE_CURRENCY_FORMAT` and `MESSAGE_TARGET_CURRENCY_FORMAT` format strings will have to change as well.

File: src/main/java/thrift/model/util/CurrencyUtil.java (hunk 5 above)
Comment (krusagiz) [reply src/main/java/thrift/model/util/CurrencyUtil.java:107]: Resolved.

**Review by lightz96 — APPROVED** [review]

File: src/main/java/thrift/logic/commands/ConvertCommand.java (hunk 6 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/commands/ConvertCommand.java]: Changed! :)

File: src/main/java/thrift/logic/parser/CliSyntax.java (hunk 7 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/parser/CliSyntax.java]: Sorted!

**Review by limerencee — APPROVED** [review]

**Pull request merged**